From 19b8e134c0d9a9654404c04ae15c958f40efa706 Mon Sep 17 00:00:00 2001
From: Pavel Emelyanov <xemul@openvz.org>
Date: Fri, 3 Apr 2009 17:13:14 +0400
Subject: [PATCH 79/82] ptrace: ban ptracing of a container init from inside the container

Current ptrace engine suffers from strange problems, one of which
is described in bug #1222 - init results in T state after incorrect
tracer detach.

Fixing it is not that easy, but since ptracing init was never alowed
before it's OK to ban this (for a while?).

http://bugzilla.openvz.org/show_bug.cgi?id=1222

Signed-off-by: Pavel Emelyanov <xemul@openvz.org>
---
 kernel/ptrace.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/kernel/ptrace.c b/kernel/ptrace.c
index df8f075..2db201b 100644
--- a/kernel/ptrace.c
+++ b/kernel/ptrace.c
@@ -530,6 +530,10 @@ struct task_struct *ptrace_get_task_struct(pid_t pid)
 {
 	struct task_struct *child;
 
+	/* ptracing of init from inside CT is dangerous */
+	if (pid == 1 && !capable(CAP_SYS_ADMIN))
+		return ERR_PTR(-EPERM);
+
 	read_lock(&tasklist_lock);
 	child = find_task_by_vpid(pid);
 	if (child)
-- 
1.6.2.2

